lib/elixir/lib/regex.ex

```diff
@@ -649,19 +649,7 @@ defmodule Regex do
 
   """
   @spec replace(t, String.t(), String.t() | (... -> String.t()), [term]) :: String.t()
-  def replace(regex, string, replacement, options \\ [])
-
-  def replace(regex, string, replacement, options)
-      when is_binary(string) and is_binary(replacement) and is_list(options) do
-    do_replace(regex, string, replacement, options)
-  end
-
-  def replace(regex, string, replacement, options)
-      when is_binary(string) and is_function(replacement) and is_list(options) do
-    do_replace(regex, string, replacement, options)
-  end
-
-  defp do_replace(%Regex{} = regex, string, replacement, options) do
+  def replace(%Regex{} = regex, string, replacement, options \\ []) do
```

Comment:
```suggestion
  def replace(%Regex{} = regex, string, replacement, options \\ [])
      when is_binary(string) and is_list(options) do
```

Reply:
@josevalim what are the advantages of adding guards on `string` and `options` when there is only one function signature?

Reply:
Better error messages if you give a bad arg!

Reply:
This is what I though, based on the _new_ compiler check added in `1.11`!

Reply:
The compiler checks help who are writing the code but not those calling the functions. The improve error message for callers was added on v1.5 IIRC. :)

Reply:
TRUE! Would it be considered a community convention/guideline to add more guards to "public" APIs, especially the ones from in the stdlib?

I’m asking more as a consistency guideline. Since many other public functions of the Regex module are not defined with guards.